cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/validation/ValidationFailure.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.etl.api.validation;
+
+import io.cdap.cdap.api.annotation.Beta;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+/**
+ * Represents a failure condition occurred during validation.
+ */
+@Beta
+public class ValidationFailure {
+  private final String message;
+  private final String correctiveAction;
+  private final List<Cause> causes;
+
+  /**
+   * Creates a validation failure with provided message.
+   *
+   * @param message validation failure message
+   */
+  public ValidationFailure(String message) {
+    this(message, null);
+  }
+
+  /**
+   * Creates a validation failure with provided message and corrective action.
+   *
+   * @param message validation failure message
+   * @param correctiveAction corrective action
+   */
+  public ValidationFailure(String message, @Nullable String correctiveAction) {
+    this.message = message;
+    this.correctiveAction = correctiveAction;
+    this.causes = new ArrayList<>();
+  }
+
+  /**
+   * Adds provided cause to this validation failure.
+   *
+   * @param cause cause of validation failure
+   * @return validation failure with provided cause
+   */
+  public ValidationFailure withCause(Cause cause) {
```

Comment:
Instead of modifying the causes, I think it's better to have a separate Builder class and make ValidationFailure immutable.